### PR TITLE
sdn: move from EventQueue to DeltaFIFO

### PR DESCRIPTION
**Pull request by dcbw**
The OpenShift controller cache should not be used in new code and is eventually going away.

See https://github.com/openshift/origin/pull/9981#discussion_r71818202

@openshift/networking @danwinship @pravisankar @smarterclayton

### CODE REVIEW

File: pkg/sdn/plugin/eventqueue.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	"k8s.io/kubernetes/pkg/client/cache"
+)
+
+func NewEventQueue(keyFunc cache.KeyFunc) *cache.DeltaFIFO {
+	return cache.NewDeltaFIFO(
+		keyFunc,
+		cache.DeltaCompressorFunc(func(d cache.Deltas) cache.Deltas {
+			return deltaCompressor(d, keyFunc)
+		}),
+		cache.NewStore(keyFunc),
```
Comment (danwinship) [inline pkg/sdn/plugin/eventqueue.go]: This doesn't seem like it should work; the last argument to NewDeltaFIFO() is treated read-only by DeltaFIFO, so by passing an empty Store (which will always remain an empty Store) you are effectively claiming that you never know about any objects, even after being told about them. This should cause you to lose Delete notifications.

File: pkg/sdn/plugin/eventqueue.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
+package plugin
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+)
+
+func NewEventQueue(keyFunc cache.KeyFunc) *cache.DeltaFIFO {
```
Comment (danwinship) [inline pkg/sdn/plugin/eventqueue.go]: it seems unlikely that the sdn code is going to be the only part of origin that wants this delta-compressing code...

File: pkg/sdn/plugin/eventqueue.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		cache.NewStore(keyFunc),
+	)
+}
+
+// Describes the effect of processing a cache event on the delta list.
+type cacheEventEffect string
+
+const (
+	// The cache event should result in an add
+	cacheEventEffectAdd cacheEventEffect = "ADD"
```
Comment (danwinship) [inline pkg/sdn/plugin/eventqueue.go]: The naming here is confusing. I think the internal delta list management needs to use different terms than the external event queue (to make it clear that this is about adding an item to the list of notifications, and not necessarily about an item being added to the set the caller is monitoring).

File: pkg/sdn/plugin/eventqueue.go (hunk 2 above)
Comment (dcbw) [inline pkg/sdn/plugin/eventqueue.go]: Currently it's just pkg/router/controller/factory/factory.go and that's apparently going to stop using it.  @smarterclayton should I just replace pkg/client/cache/eventqueue.go with this new DeltaFIFO-based event queue, and convert everything over to use it?

**Comment by dcbw** [comment]
@danwinship I fixed for all your comments, now repushed.  I'd like to get this in earlier since it's blocking the CNI plugin stuff, and then perhaps we could make it more generalized and move the router controller factory over to it later?

**Comment by dcbw** [comment]
[test]

File: pkg/sdn/plugin/eventqueue.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/sdn/plugin/eventqueue.go]: It's only router and SDN, so if you're confident in the correctness of your implementation and we have enough time / effort to properly test it, then I'm ok to go ahead there.

**Comment by smarterclayton** [comment]
Changes look good other than the comment about replacing the router code as well.

**Comment by dcbw** [comment]
> Changes look good other than the comment about replacing the router code as well.

@smarterclayton thanks; is it your intention to move the router code away from NewEventQueue, or just keep it private to the router code?

**Comment by pravisankar** [comment]
LGTM

**Comment by smarterclayton** [comment]
Goa is to replace router with an informer + indexer and use the events
to trigger recalculation.  However, needs to be worked through in more
detail.

**Comment by dcbw** [comment]
@openshift/networking @danwinship @pravisankar back up for review; but we may want to *not* merge this until the freeze is done.

**Comment by dcbw** [comment]
Flake:

```
++ Building go targets for linux/amd64: cmd/openshift cmd/oc

# github.com/openshift/origin/vendor/google.golang.org/grpc/transport

vendor/google.golang.org/grpc/transport/http_util.go:271: internal compiler error: schedule does not include all values
```

**Comment by dcbw** [comment]
[test]

**Comment by dcbw** [comment]
flake is https://github.com/openshift/origin/issues/10445

**Comment by dcbw** [comment]
re-[test]

**Comment by dcbw** [comment]
re-[test] flake #10445

**Comment by dcbw** [comment]
re-[test] now that travis seems happier

**Comment by dcbw** [comment]
re-[test] flake https://github.com/openshift/origin/issues/9938

**Comment by dcbw** [comment]
Updated for a bug; Deleted events wouldn't be propagated when the queue was empty for an object, due to how DeltaFIFO tracks things internally when knownObjects is nil.  Thus we must implement our own store in EventQueue to get reliable deletion notification.

**Comment by dcbw** [comment]
jenkins failure is flake https://github.com/openshift/origin/issues/10080

**Comment by dcbw** [comment]
[test]

**Comment by danwinship** [comment]
[merge]

**Pull request merged**